### DISCUSSION TITLE
many-to-many relationships: get child Ids without querying the child table, just the association table?

### DISCUSSION THREAD

**Discussion by spolloni** [sqlalchemy/sqlalchemy#7859] [Usage Questions] [answered]
My setup resembles the following:

```
from sqlalchemy_utils import create_view

association_table = Table('association', Base.metadata,
    Column('gambler_id', ForeignKey('gamblers.id')),
    Column('horse_id', ForeignKey('horses.id'))
)

class Gambler(Base):
    __tablename__ = 'gamblers'
    id = Column(Integer, primary_key=True)
    horses = relationship("Horse",
                    secondary=association_table,
                    primary_join=...,
                    secondary_join=...)

some_selectable = select(...) # there is an 'id' column here 

class Horse(Base):
    __table__ = create_view('horses', some_selectable, Base.metadata)
```

Notice this is essentially the docs many-to-many [example](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-many), with the distinction that the horses table is actually a VIEW, which itself is a `UNION ALL` of several other tables. Because of this, I need to provide explicit statement in the `primary_join` and `secondary_join` args. We can think of the set of horses as being fixed (no writes), but the set of gamblers and the horses they bet on to change a bunch.  

Now I do realize using a VIEW may be a questionable DB design, but this has so far avoided other headaches and worked pretty well. However, I recently added new columns to the horses VIEW, and this has degraded performance **significantly**. 

My use case is one where I need to know often and quickly about which horses a gambler is betting on, but can afford the perfomance hit when needing other attributes of those horses. I am wondering if there is a way to setup the `relationship` so that the VIEW is not queried, just the association table, to get the `horse_id`s for a given `Horse`.

My current attempt is this: 

```
class Gambler(Base):
    __tablename__ = 'gamblers'
    id = Column(Integer, primary_key=True)
    horses = relationship("Horse",
                    secondary=association_table,
                    primary_join=...,
                    secondary_join=...)
    _horses = relationship(
                    Mapper(
                        type('GamblerHorse', (), {}),
                        association_table,
                        primary_key=[
                            association_table.c.gambler_id,
                            association_table.c.horse_id
                        ]
                    )
                    view_only=True)
    horse_ids = association_proxy('_horses', 'horse_id')
```

but am wondering if there is a better/easier implementation.

**Comment by CaselIT** [comment]
Hi,

You can transform `association_table` in an orm model, then setup a viewonly relationship from gambler to that model.

**Reply by zzzeek** [reply]
that seems to be what they are already doing when they use Mapper().

**Comment by zzzeek** [comment, marked as answer]
if an ORM-managed collection of scalar values is the effect that you want, then yes the association proxy against a class mapped to the table that contains the scalar values is how you would do this.

as far as other solutions to the issue, I would need to see the nature of this view and what it's trying to query.

**Reply by spolloni** [reply]
> yes the association proxy against a class mapped to the table that contains the scalar values is how you would do this.

thanks @zzzeek ! We did keep that pattern in the end. 

We also fixed the view's performance by removing "derivative" columns in the `VIEW` definition that didn't physically exist in the set of tables that were being `UNION`ed. Instead we implemented those columns with `column_property()`
